src/utils/resource-container-helper.js

```diff
@@ -0,0 +1,146 @@
+const DEEP = true;
+
+const ensureLoaded = (element) => element.ensureLoaded();
+const pause = (element) => element.pause();
+const unmount = (element) => element.unmount();
+
+/**
+ * Ensure all elements within this container are scheduled to load.
+ *
+ * @param {!Element|!Array<!Element>} containerOrContainers
+ * @param {boolean=} includeSelf
+ */
+export function loadAll(containerOrContainers, includeSelf = true) {
+  forAllWithin(containerOrContainers, includeSelf, !DEEP, ensureLoaded);
+}
+
+/**
+ * Pause all elements within this container.
+ *
+ * @param {!Element|!Array<!Element>} containerOrContainers
+ * @param {boolean=} includeSelf
+ */
+export function pauseAll(containerOrContainers, includeSelf = true) {
+  forAllWithin(containerOrContainers, includeSelf, DEEP, pause);
+}
+
+/**
+ * Unmount all elements within this container.
+ *
+ * @param {!Element|!Array<!Element>} containerOrContainers
+ * @param {boolean=} includeSelf
+ */
+export function unmountAll(containerOrContainers, includeSelf = true) {
+  forAllWithin(containerOrContainers, includeSelf, DEEP, unmount);
+}
+
+/**
+ * Execute a callback for all elements within the container.
+ *
+ * @param {!Element|!Array<!Element>} containerOrContainers
+ * @param {boolean} includeSelf
+ * @param {boolean} deep
+ * @param {function(!AmpElement)} callback
+ */
+export function forAllWithin(
+  containerOrContainers,
+  includeSelf,
+  deep,
+  callback
+) {
+  if (Array.isArray(containerOrContainers)) {
+    for (let i = 0; i < containerOrContainers.length; i++) {
+      forAllWithinInternal(
+        containerOrContainers[i],
+        includeSelf,
+        deep,
+        callback
+      );
+    }
+  } else {
+    forAllWithinInternal(containerOrContainers, includeSelf, deep, callback);
+  }
+}
+
+/**
+ * Execute a callback for all elements within the container.
+ *
+ * @param {!Element} container
+ * @param {boolean} includeSelf
+ * @param {boolean} deep
+ * @param {function(!AmpElement)} callback
+ */
+function forAllWithinInternal(container, includeSelf, deep, callback) {
+  if (includeSelf && container.classList.contains(AMP_CLASS)) {
+    callCallback(callback, container);
+    if (!deep) {
+      // Also schedule amp-element that is a placeholder for the element.
+      const placeholder = container.getPlaceholder();
+      if (placeholder) {
+        forAllWithinInternal(
+          placeholder,
+          /* includeSelf */ true,
+          !DEEP,
+          callback
+        );
+      }
+      return;
+    }
+  }
+
+  const descendants = container.getElementsByClassName(AMP_CLASS);
+  let seen = null;
+  for (let i = 0; i < descendants.length; i++) {
+    const descendant = descendants[i];
+    if (deep) {
+      // In deep search all elements will be covered.
+      callCallback(callback, descendant);
+    } else {
+      // Breadth-first search. Rely on the `getElementsByClassName` DOM order
+      // to ignore DOM subtrees already covered.
+      seen = seen || [];
```

Comment:
should this be a Set?

Reply:
Copied it from `owners-impl.js`. But the reason it's not a set is because the test is not `seen.contains(target)`, but `seen.some(x => x.contains(target))`. The uniqueness of `seen` itself is guaranteed by the `querySelectorAll`. Do you think those cases are still improved by using a set?

Reply:
No, iterating a Set is considerably slower than iterating an array. Given we must do a `some`, the Array is the right choice.

Reply:
Yeah. That's what I figured as well.